### ISSUE TITLE
How to interact with systray with keyboard

### ISSUE THREAD

**Issue by Bew78LesellB** [awesomeWM/awesome#212]
Hello,

I have an app that is usable with the keyboard once it's focused. When we try to close it, it goes in the systray.. Now, how can I reopen it, without using the mouse ? (a click on the systray icon toggle the app visibility)

I tried to look in `/usr/share/awesome/lib/wibox/widget/systray.lua` but I didn't find any way to control the systray..

So, is there a way ? how ?
Or can it be added in next version ?

Thanks
Bew

**Comment by syphoxy** [comment]
I was actually playing with rodentbane today. it's a library that adds an interface for navigating with the mouse but with your keyboard (by using a grid system). that being said, you can use rodentbane for this but I feel like we could go a step further and add an easy interface for warping the mouse and clicking/right clicking on systray icons.

**Comment by Bew78LesellB** [comment]
Yes thanks for the idea. I'll look into that library. 
And yes I was looking for the easy interface on the systray.

I'd like to do the modifications for this to work, but I don't know where to start, do I need to change some C code?

**Comment by psychon** [comment]
If you want to simulate a click on some systray icon, you can do that from lua (almost):
* Figure out the systray position on the screen (this might be the hard part)
* Calculate the position of the systray icon that you want to simulate a click on.
* Warp the mouse to that position (I think `mouse.coords` allows to do that)
* Use `root.fake_input` to fake a mouse click

I guess the C code would have to do almost the same.
However, it could fake a mouse click directly without warping the mouse by faking the ButtonPress and ButtonRelease events directly, but the program behind the systray icon could detect this. Dunno if anything would care about such a fake...

**Comment by Bew78LesellB** [comment]
When you say `Calculate the position of the systray icon that you want to simulate a click on.` :
How to know which app in the systray icon list we are looking for? 


Is there a way to get the systray infos, something like an array of objects for each app, containing :
- the systray icon
- a unique name or id

So on the lua side we could make our own systray widget for example...

**Comment by psychon** [comment]
Nope, all that the C code hands to lua is the number of entries in the systray.

**Comment by Bew78LesellB** [comment]
I think we need to change this to be able to interact with it...

**Comment by syphoxy** [comment]
to be fair, the number of items is actually all you need. the rest of it can be assumed as systray icons are always square. as long as you can grab the systray widget height, you can start creating a mouseless method to navigate the items.

for example, to get the mouse centered on the second item (*warning: pseudo code*):

```
h = systray.height

# assuming top/left point of widget
x = systray.position.x 
y = systray.position.y

target_entry = 2

# assuming horiziontal systray
warp_cursor((x + h * target_entry) - (h / 2), y + height / 2)
```

something like this?

**Comment by Bew78LesellB** [comment]
I see, right, but here (maybe it's a bit too much) we can't SET that the app A systray's icon will be the second, because if I open another app B, the app A will be the 3th not the 2nd...

But that maybe too much (and useless to handle)..

**Comment by osleg** [comment]
rodentbane does not draw itself over the wibox, you can move it with shift+direction but it'll add to complexity.
On the other hand it could be possible to take rodentbane, make it draw simplified grid with only 3 vertically split boxes over the wibox and navigate from there.

**Comment by syphoxy** [comment]
yep. something like that. :)

**Comment by osleg** [comment]
Already working on it ^_^
